lib/Runtime/HostManager/HostManager.cpp

```diff
@@ -334,47 +368,43 @@ HostManager::runNetwork(llvm::StringRef networkName,
       network = &it->second;
       network->refcount++;
     }
-  }
 
-  if (network == nullptr) {
-    callback(
-        currentRun,
-        MAKE_ERR(GlowErr::ErrorCode::RUNTIME_NET_NOT_FOUND,
-                 llvm::formatv("Function {0} not found", networkName).str()),
-        std::move(context));
-    return currentRun;
+    if (network == nullptr) {
+      callback(
+          currentRun,
+          MAKE_ERR(GlowErr::ErrorCode::RUNTIME_NET_NOT_FOUND,
+                   llvm::formatv("Function {0} not found", networkName).str()),
+          std::move(context));
+      return currentRun;
+    }
+    // Setup the request
+    InferRequest queuedRequest(networkName, std::move(context), callback,
+                               priority, currentRun);
+    // Put the request in the queue.
+    auto queueSize = inferQueue_.size();
+    if (queueSize >= config_.maxQueueSize) {
+      // The queue is full, return an error.
+      network->refcount--;
+      callback(
+          currentRun,
+          MAKE_ERR(
+              GlowErr::ErrorCode::RUNTIME_REQUEST_REFUSED,
+              strFormat(
+                  "The number of allowed queued requests has been exceeded. "
+                  "queued requests: %lu allowed requests: %zu",
+                  queueSize, config_.maxQueueSize)),
+          std::move(context));
+      return currentRun;
+    }
+    inferQueue_.push(std::move(queuedRequest));
```

Comment:
does not it race with inferQueue_.pop() https://github.com/pytorch/glow/pull/3416/files#diff-0ff368c87ec74b6b37e29804c4122427R327 ? (It seems that we do push without any lock here).

Reply:
There's a lock on L365 it's just waaaay up the page (hopefully I'm counting braces correctly :-p )

Reply:
ah, right '}' was removed and expand that lock to larger scope.

Reply:
hmm yeah the push should be under a lock.

Reply:
one more dumb comment while we are here :)

Invoking callback under the lock might be a risky business ...

Reply:
That's an interesting point, given we don't know what the callback does, it may not be totally safe to hold the lock through it.  @gcatron can you follow up?